atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/limiter/ClientLimiterImpl.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra.limiter;
+
+import com.palantir.atlasdb.keyvalue.cassandra.limiter.SafeSemaphore.CloseablePermit;
+import java.io.IOException;
+import org.apache.cassandra.thrift.TimedOutException;
+import org.apache.thrift.TException;
+
+public class ClientLimiterImpl implements ClientLimiter {
+
+    private final SafeSemaphore<TException> concurrentRangeScans;
+
+    public ClientLimiterImpl(int maxConcurrentRangeScans) {
+        concurrentRangeScans = SafeSemaphore.of(maxConcurrentRangeScans, TimedOutException::new);
```

Comment:
TimedOutException?

> either a node failed mid-operation, or load was too high, or the requested op was too large

abuse of this exception, maybe?

'load was too high'

I can also make some custom exception for this - thoughts?

Reply:
Given the goal is to reduce usage of range scans in general, it seems like any non-retryable exception should work. More descriptive might be better but don't have a strong opinion on whether to reuse an existing or create a new exception.